src/main/java/spoon/reflect/visitor/chain/CtQueryListener.java

```diff
@@ -0,0 +1,41 @@
+/**
+ * Copyright (C) 2006-2017 INRIA and contributors
+ * Spoon - http://spoon.gforge.inria.fr/
+ *
+ * This software is governed by the CeCILL-C License under French law and
+ * abiding by the rules of distribution of free software. You can use, modify
+ * and/or redistribute the software under the terms of the CeCILL-C license as
+ * circulated by CEA, CNRS and INRIA at http://www.cecill.info.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the CeCILL-C License for more details.
+ *
+ * The fact that you are presently reading this means that you have had
+ * knowledge of the CeCILL-C license and that you accept its terms.
+ */
+package spoon.reflect.visitor.chain;
+
+/**
+ * Listens for entering/exiting of query step
+ */
+public interface CtQueryListener {
+	/**
+	 * Called when new object enters query step, before the object is processed by mapping function of that step.
+	 * <br>
+	 * The returning false or throwing an exception causes that processing of input object and all it's children is skipped.
+	 * The returning true causes that input object is normally processed, and exit is called for that object.
+	 * @param element the processed element
+	 * @param context context the query execution context
+	 * @return true to continue processing this step and it's children,
+	 * false to skip this step and it's children.
+	 */
+	boolean enter(Object element, CtQuery context);
+	/**
+	 * The exit is called always only after `enter` returns true and input element and all children are processed or they throw an exception.
+	 * @param element the processed element
+	 * @param context the query execution context
+	 * @param exception null or an exception if the processing of the mapping function failed

```

Comment:
to detect whether processing of the children passed correctly or failed. I do not need it now, because my code does not throw exceptions, but it is API so it should be generic enough